### DISCUSSION TITLE
Feature request (?): support two-step authentication

### DISCUSSION THREAD

**Discussion by danvinci** [fonsp/Pluto.jl#908] [Pluto development]
Ok, not sure the title is correct to begin with - I'll explain what I mean:

I mostly use Pluto as a learning environment, and often try to replicate material published on Jupyter notebooks.
It's often the case that some form of authentication is necessary in order to get data from a 3rd-party, at least this is the case with Google services like Earth Engine.

For example, the following Python code, in a Jupyter notebook:

```
import ee
ee.Authenticate() 
ee.Initialize()
```
Results in something like this, and expects the user to input the verification code:
![Screenshot 2021-02-09 at 1 09 11 PM](https://user-images.githubusercontent.com/15657499/107361666-05570000-6ad8-11eb-8057-f9aa3cc8012c.png)

In Pluto:

```
begin
   using PyCall
   ee = pyimport("ee")
   ee.Authenticate() 
   ee.Initialize()
end
```

Will give the same output shown above in the REPL (still totally fine), but the REPL seems not responsive to any input other than ctrl+c or ctrl+d, so pasting the verification code there won't yeld any result.

The process then goes on indefinitely, waiting for the confirmation code, and when interrupting it I get a Distributed.ProcessExitedException and Pluto must be restarted after editing/commenting the line that caused the issue.

Some features being considered would surely help having a 'graceful 'handling of execution issues, like starting notebooks as frozen and/or having an execution barrier - but my question is:

would it be worth supporting two-step authentication workflows like these, as a case for feature-parity with Jupyter?
Or am I running into a niche/edge case specific to Google?

On a side note: I'd be using Python just for interfacing with EE, while doing everything else with Julia once I get the data. There's a beta REST API that would be worth exploring as the authentication would be handled through a key exchange (it seems to me), but at the moment doesn't cover the same scope of the Python API. 

Is anyone else using Pluto for overlapping use cases? (e.g. pulling data from sources requiring some form of auth)

**Comment by fonsp** [comment]
You can definitely use Pluto to make requests with authentication, because you can run arbitrary Julia code! Making web requests with 2FA is not a feature that an IDE needs to provide.

Your question is actually much more specific: you tried to use the python `ee` package through PyCall inside Pluto, which calls `readline` as an input method, which is not working. For that, take a look at https://github.com/fonsp/Pluto.jl/discussions/765

**Reply by fonsp** [reply]
Also note that `feature-parity with Jupyter` is not one of our design goals!
